src/main/kotlin/org/jetbrains/kotlinx/jupyter/connection.kt

```diff
@@ -150,14 +227,208 @@ class JupyterConnection(val config: KernelConfig) : Closeable {
+    val stdin = Socket(JupyterSocketInfo.STDIN)
+    val iopub = Socket(JupyterSocketInfo.IOPUB)
+
+    private fun fromSocketName(socket: JupyterSocket): Socket {
+        return when (socket) {
+            JupyterSocket.HB -> heartbeat
+            JupyterSocket.SHELL -> shell
+            JupyterSocket.CONTROL -> control
+            JupyterSocket.STDIN -> stdin
+            JupyterSocket.IOPUB -> iopub
+        }
+    }
+
+    private val callbacks = mutableMapOf<RawMessageCallback, SocketRawMessageCallback>()
+
+    override fun addMessageCallback(callback: RawMessageCallback): RawMessageCallback {
+        val socket = fromSocketName(callback.socket)
+        val socketCallback: SocketRawMessageCallback = { message ->
+            if (message.type == callback.messageType) {
+                callback.action(message)
+            }
+        }
+        callbacks[callback] = socket.onRawMessage(socketCallback)
+        return callback
+    }
+
+    override fun removeMessageCallback(callback: RawMessageCallback) {
+        val socketCallback = callbacks[callback] ?: return
+        val socket = fromSocketName(callback.socket)
+        socket.removeCallback(socketCallback)
+    }
+
+    fun updateSessionInfo(message: Message) {
+        val header = message.data.header ?: return
+        header.session?.let { sessionId = it }
+        header.username?.let { username = it }
+        messageId = message.id
+    }
+
+    override fun send(socketName: JupyterSocket, message: RawMessage) {
+        val socket = fromSocketName(socketName)
+        socket.sendRawMessage(message)
+    }
+
+    override fun sendReply(socketName: JupyterSocket, parentMessage: RawMessage, type: String, content: JsonObject, metadata: JsonObject?) {
+        val message = RawMessageImpl(
+            parentMessage.id,
+            JsonObject(
+                mapOf(
+                    "header" to makeJsonHeader(type, parentMessage),
+                    "parent_header" to (parentMessage.header ?: Json.EMPTY),
+                    "metadata" to (metadata ?: Json.EMPTY),
+                    "content" to content
+                )
+            )
+        )
+        send(socketName, message)
+    }
+
+    private val commOpenCallbacks = ConcurrentHashMap<String, CommOpenCallback>()
+    private val commTargetToIds = ConcurrentHashMap<String, CopyOnWriteArrayList<String>>()
+    private val commIdToComm = ConcurrentHashMap<String, CommImpl>()
+    override fun openComm(target: String, data: JsonObject): Comm {
+        val id = UUID.randomUUID().toString()
+        val newComm = processCommOpen(target, id, data)
+
+        // send comm_open
+        iopub.sendMessage(
+            makeSimpleMessage(
+                MessageType.COMM_OPEN,
+                CommOpen(newComm.id, newComm.target)
+            )
+        )
+
+        return newComm
+    }
+
+    fun processCommOpen(target: String, id: String, data: JsonObject): Comm {
```

Comment:
Here we assume that registration is always successful, but if there is an exception, we do not do anything. According to [doc](https://jupyter-client.readthedocs.io/en/latest/messaging.html#opening-a-comm), if it fails we actually might send a commClose to notify failure